### PR TITLE
Add sandbox mode to exchange

### PR DESCRIPTION
**Pull request by stevenewald**

### CODE REVIEW

**Review by echavemann — APPROVED** [review]
Fine with this functionally, I'd expect Nino to have some style remarks

File: exchange/CMakeLists.txt [hunk 1]
```diff
@@ -83,6 +89,10 @@ target_link_libraries(NUTC24_exe PRIVATE quill::quill)
 # fmt
 target_link_libraries(NUTC24_exe PRIVATE fmt::fmt)
 
+# libzip
+find_package(libzip REQUIRED)
```
Comment (egelja) [inline exchange/CMakeLists.txt:93]: dont' call this twice

File: exchange/src/client_manager/client_manager.hpp [hunk 2]
```diff
@@ -30,7 +31,8 @@ class ClientManager {
 
 public:
     void add_client(
-        const std::string& uid, float capital = STARTING_CAPITAL, bool active = false
+        const std::string& uid, float capital = STARTING_CAPITAL,
+        bool is_local_algo = false, bool active = false
```
Comment (egelja) [inline exchange/src/client_manager/client_manager.hpp:35]: Don't do default arguments. What if I want to specify `is_local_algo` but not `capital`?

Just use function overloading.

File: exchange/src/curl/curl.cpp [hunk 3]
```diff
@@ -15,8 +17,55 @@ write_callback(void* contents, size_t size, size_t nmemb, void* userp)
...
+{
+    CURL* curl;
+    CURLcode res;
+    FILE* fp;
+
+    curl = curl_easy_init();
+    if (curl) {
+        fp = fopen(filepath.c_str(), "wb");
+        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
+        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, file_write_callback);
```
Comment (egelja) [inline exchange/src/curl/curl.cpp:40]:
```suggestion
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, fwrite);
```

No need to wrap `fwrite` if you just pass it.

File: exchange/src/curl/curl.cpp [hunk 4]
```diff
@@ -15,8 +17,55 @@ write_callback(void* contents, size_t size, size_t nmemb, void* userp)
...
+    const std::string& data
+)
+{
+    CURL* curl;
+    CURLcode res;
+    FILE* fp;
+
+    curl = curl_easy_init();
+    if (curl) {
+        fp = fopen(filepath.c_str(), "wb");
```
Comment (egelja) [inline exchange/src/curl/curl.cpp:38]: You should check this return code the file opening may fail

File: exchange/src/curl/curl.hpp [hunk 5]
```diff
@@ -11,7 +10,17 @@ namespace nutc {
 /**
  * @brief Handles all communication with firebase
  */
-namespace firebase {
+namespace curl {
+
+// TODO: better name
+std::string request_to_string(
+    const std::string& method, const std::string& url, const std::string& data = ""
```
Comment (egelja) [inline exchange/src/curl/curl.hpp:17]: Why not make `data` a `std::optional`?

File: exchange/src/curl/curl.hpp [hunk 6]
```diff
@@ -11,7 +10,17 @@ namespace nutc {
...
+);
+
+void request_to_file(
+    const std::string& method, const std::string& url, const std::string& filename,
+    const std::string& data = ""
```
Comment (egelja) [inline exchange/src/curl/curl.hpp:22]: Why not make `data` a `std::optional`?

File: exchange/src/logging.hpp [hunk 7]
```diff
@@ -82,6 +82,7 @@ CREATE_LOG_CATEGORY(client_spawning);
 CREATE_LOG_CATEGORY(rabbitmq);
 CREATE_LOG_CATEGORY(dev_mode);
 CREATE_LOG_CATEGORY(events);
+CREATE_LOG_CATEGORY(sandbox_mode);
```
Comment (egelja) [inline exchange/src/logging.hpp:85]:
```suggestion
CREATE_LOG_CATEGORY(sandbox);
```

File: exchange/src/process_spawning/spawning.cpp [hunk 8]
```diff
@@ -1,28 +1,34 @@
...
 
 namespace nutc {
 namespace client {
 
 int
-initialize(manager::ClientManager& users, bool development_mode)
+initialize(manager::ClientManager& users, Mode mode)
 {
-    if (development_mode) {
+    if (mode == Mode::DEV) {
```
Comment (egelja) [inline exchange/src/process_spawning/spawning.cpp:15]: Make this a switch statement

File: exchange/src/utils/local_algos/dev_mode.cpp [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+void
+create_mt_algo_files(int num_users) noexcept
+{
+    std::string content = file_mgmt::read_file_content("./template.py");
+    std::string dir_name = std::string(ALGO_DIR);
+    if (!file_mgmt::create_directory(dir_name))
+        log_e(dev_mode, "{}", "Failed to create directory.");
+
+    for (int i = 0; i < num_users; i++) {
+        std::string file_name = dir_name + "/algo_" + std::to_string(i) + ".py";
```
Comment (egelja) [inline exchange/src/utils/local_algos/dev_mode.cpp:28]: I'd recommend `std::filesystem::path` but this should be fine too

File: exchange/src/utils/local_algos/dev_mode.cpp [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+        std::string uid = "algo_" + std::to_string(i);
+        users.add_client(uid, STARTING_CAPITAL, true);
+    }
+}
+
+void
+create_mt_algo_files(int num_users) noexcept
+{
+    std::string content = file_mgmt::read_file_content("./template.py");
+    std::string dir_name = std::string(ALGO_DIR);
```
Comment (egelja) [inline exchange/src/utils/local_algos/dev_mode.cpp:23]: I'd recommend `std::filesystem::path` but this should be fine too

File: exchange/src/utils/local_algos/file_management.cpp [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    }
+    return true;
+}
+
+bool
+file_exists(const std::string& filename) noexcept
+{
+    std::ifstream file(filename);
+    return file.good();
+}
```
Comment (egelja) [inline exchange/src/utils/local_algos/file_management.cpp:82]: `std::filesystem::path`

File: exchange/src/utils/local_algos/file_management.cpp [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+{
+    struct stat st;
+    if (stat(dir_name.c_str(), &st) != 0) {
+        if (mkdir(dir_name.c_str(), 0777) != 0) {
+            log_e(dev_mode, "{}", "Failed to create directory.");
+            return false;
+        }
+    }
+    return true;
+}
```
Comment (egelja) [inline exchange/src/utils/local_algos/file_management.cpp:75]: `std::filesystem::path`

Shouldn't be using C functions if needed.

File: exchange/src/utils/local_algos/file_management.cpp [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+        if (f == nullptr) {
+            log_e(dev_mode, "Error opening file in zip: {}", zip_strerror(z));
+            continue;
+        }
+
+        char* contents = new char[st.size];
+        zip_fread(f, contents, st.size);
+
+        // Construct the full path for the unzipped file
+        std::string full_path = dest + "/" + st.name;
```
Comment (egelja) [inline exchange/src/utils/local_algos/file_management.cpp:50]: `std::filesystem::path`

File: exchange/src/utils/local_algos/file_management.cpp [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+
+    zip_uint64_t num_entries = static_cast<zip_uint64_t>(zip_get_num_entries(z, 0));
+    for (zip_uint64_t i = 0; i < num_entries; i++) {
+        struct zip_stat st;
+        zip_stat_index(z, i, 0, &st);
+
+        zip_file* f = zip_fopen_index(z, i, 0);
+        if (f == nullptr) {
+            log_e(dev_mode, "Error opening file in zip: {}", zip_strerror(z));
+            continue;
```
Comment (egelja) [inline exchange/src/utils/local_algos/file_management.cpp:43]: Shouldn't you exit on error?

File: exchange/CMakeLists.txt (hunk 1 above)
Comment (stevenewald) [reply exchange/CMakeLists.txt:93]: Oop

File: exchange/src/client_manager/client_manager.hpp (hunk 2 above)
Comment (stevenewald) [reply exchange/src/client_manager/client_manager.hpp:35]: Good point

File: exchange/src/curl/curl.hpp (hunk 5 above)
Comment (stevenewald) [reply exchange/src/curl/curl.hpp:17]: Keeping as default

File: exchange/src/utils/local_algos/dev_mode.cpp (hunk 10 above)
Comment (stevenewald) [reply exchange/src/utils/local_algos/dev_mode.cpp:23]: This is just to convert algo_dir to a string

File: exchange/src/client_manager/client_manager.cpp [hunk 15]
```diff
@@ -64,13 +64,19 @@ void
...
 {
-    clients[uid] = Client{uid, active, capital, {}};
+    clients[uid] = Client{uid, false, is_local_algo, STARTING_CAPITAL, {}};
+}
+
+void
+ClientManager::add_client(const std::string& uid)
+{
+    clients[uid] = Client{uid, false, false, STARTING_CAPITAL, {}};
 }
```
Comment (egelja) [inline exchange/src/client_manager/client_manager.cpp:80]: May be good to make these inline funcs in the header

File: exchange/src/utils/local_algos/file_management.cpp [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+
+        delete[] contents;
+
+        zip_fclose(f);
+    }
+    zip_close(z);
+}
+
+bool
+create_directory(std::string dir_name)
```
Comment (egelja) [inline exchange/src/utils/local_algos/file_management.cpp:65]: Pass as reference

Comment (egelja) [inline exchange/src/utils/local_algos/file_management.cpp:65]: Pass as reference

**Review by egelja — CHANGES_REQUESTED** [review]

**Pull request merged**